python/detail.cpp

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2019 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+
+#include "scipp/units/unit.h"
+
+#include "bind_detail.h"
+#include "pybind11.h"
+
+using namespace scipp;
+using namespace scipp::core;
+
+namespace py = pybind11;
+
+void init_detail(py::module &m) {
+
+  py::class_<MoveableVariable>(m, "MoveableVariable");
+  py::class_<MoveableDataArray>(m, "MoveableDataArray");
+
+  auto detail = m.def_submodule("detail");
+
+  detail.def("move",
+             [](Variable &var) {
+               auto mv = MoveableVariable{std::move(var)};
+               var = makeVariable<bool>(Dims{Dim::X}, Shape{0});
+               return mv;
+             },
+             py::call_guard<py::gil_scoped_release>(), R"(
+        This function can be used in a similar way to the C++ std::move to
+        transfer ownership of an input Variable to a container.
+        This is useful when wanting to avoid unnecessary copies of Variables
+        when inserting them into a Dataset.
+
+        :return: A MoveableVariable wrapper class.
+        :rtype: MoveableVariable)");
+
+  detail.def("move",
+             [](DataArray &data) {
+               auto mda = MoveableDataArray{std::move(data)};
+               data = DataArray();
+               return mda;
+             },
+             py::call_guard<py::gil_scoped_release>(), R"(
+        This function can be used in a similar way to the C++ std::move to
+        transfer ownership of an input DataArray to a container.
+        This is useful when wanting to avoid unnecessary copies of Variables
+        when inserting them into a Dataset.
+
+        :return: A MoveableDataArray wrapper class.
+        :rtype: MoveableDataArray)");
+
+  detail.def(
+      "move_to_data_array",
+      [](Variable &data, std::map<Dim, Variable &> &coords,
+         std::map<std::string, Variable &> &labels,
+         std::map<std::string, Variable &> &masks,
+         std::map<std::string, Variable &> &attrs, const std::string &name) {
+        return DataArray(std::move(data), std::move(coords), std::move(labels),
+                         std::move(masks), std::move(attrs), name);
```

Comment:
In `move` you set the input to a valid but empty variable, here you do not. Would probably be complicated to do here, maybe just remove the special handling in `move` and rely on good documentation?

Reply:
yes it was complicated to do here. I'm ok with removing it from the `move`.